plugins/nf-azure/src/main/nextflow/cloud/azure/batch/AzHelper.groovy

```diff
@@ -198,4 +200,41 @@ class AzHelper {
                 .buildClient()
     }
 
+    @Memoized
+    static synchronized BlobServiceClient getOrCreateBlobServiceWithServicePrincipal(String accountName, String clientId, String clientSecret, String tenantId) {
+        log.debug "Creating Azure Blob storage client using Service Principal credentials"
+
+        final endpoint = String.format(Locale.ROOT, "https://%s.blob.core.windows.net", accountName);
+
+        final servicePrincipalBasedCred = new ClientSecretCredentialBuilder()
+                .clientId(clientId)
+                .clientSecret(clientSecret)
+                .tenantId(tenantId)
+                .build()
+
+        return new BlobServiceClientBuilder()
+                .credential(servicePrincipalBasedCred)
+                .endpoint(endpoint)
+                .buildClient()
+    }
+
+    @Memoized
+    static synchronized BlobServiceClient getOrCreateBlobServiceWithManagedIdentity(String accountName, String clientId) {
+        if( !clientId )
+            throw new IllegalArgumentException("Missing Azure blob managed identity client ID")
+
+        log.debug "Creating Azure blob storage client -- accountName: $accountName; clientId: ${clientId}"
+
+        final endpoint = String.format(Locale.ROOT, "https://%s.blob.core.windows.net", accountName)
+
+        final credential = new DefaultAzureCredentialBuilder()
+                .managedIdentityClientId(clientId)
+                .build()
```

Comment:
We'll need another config option to explicitly enable the system assigned identity.

Here are the docs for the default credential builder: https://learn.microsoft.com/en-us/java/api/com.azure.identity.defaultazurecredentialbuilder?view=azure-java-stable

It's not obvious to me that it defaults to the system-assigned identity, since it seems to support other credentials. But maybe you can discern better than me.

Reply:
Does this mean that with the system-assigned identity you would only need to know the batch/storage account names and then you could submit jobs from anywhere without any client-side credentials?

Reply:
Or is the "system" the node from which you are submitting tasks i.e. the Nextflow head job which would presumably be running in Azure?